exec/cnex/number.c

```diff
@@ -17,49 +17,88 @@
 char *number_to_string(Number x)
 {
     static char buf[50];
-    char val[50];
-    bid128_to_string(val, x);
-    sprintf(buf, "%g", atof(val));
+    number_toString(x, buf, sizeof(buf));
     return buf;
 }
 
-/* ToDo: Implement proper number to string formatter. */
 void number_toString(Number x, char *buf, size_t len)
 {
-    //const int PRECISION = 34; /* ToDo: Finish implementation */
-    int iLeadingZeros = 0;
+    memset(buf, 0x00, len);
+    const int PRECISION = 34;
 
     assert(len != 0);
 
-    char val[50];
-    char lead[50];
+    char val[50] = { 0 };
     bid128_to_string(val, x);
 
-    char *s = &val[0];
-
-    /* Skip over the leading +, if there is one. */
-    while(*s != '\0' && (*s == '+')) s++;
-
-    char *v = s;
+    char *v, *s = v = &val[1];
 
-    while (*s != '\0' && (*s != 'E')) s++;
+    // Calculate the length of the string, locating the exponent value at the same time.
+    int slen = 0;
+    while (*s != '\0' && (*s != 'E')) s++, slen++;
```

Comment:
Please don't do this stuff where you cram multiple things into one statement on one line. Use braces. Use multiple lines. Use explanatory variable names (not single letters). Make it clear.

Reply:
Fixed.